docs/MULTIMODELSERVING_GUIDE.md

```diff
@@ -1,29 +1,98 @@
-Additionally, there is no easy way to request a fraction of GPU in Kubernetes infrastructure, it makes sense to load multiple models in one model server to share GPU resources. KFServing's multi-model serving is a solution that allows for loading multiple models into a server while still keeping the out of the box serverless features.
+The original design of KFServing deploys one model per InferenceService. 
+But, its 'one model, one server' paradigm presents challenges for a Kubernetes cluster when dealing with a large number of models. 
+To scale the number of models, we have to scale the number of InferenceServices, 
+something that can quickly challenge the cluster's limits.
 
-### Benefits
-- Allow multiple models to share the same GPU
-- Increase the total number of models that can be deployed in a cluster
-- Reduced model deployment resource overhead
-    - An InferenceService needs some CPU and overhead for each replica
-    - Loading multiple models in one inferenceService is more resource efficient
-    - Allow deploying hundreds of thousands of models with ease and monitoring deployed trained models at scale
+Multi-model serving is designed to address three types of limitations KFServing will run into: 1) Compute resource limitation, 2) Maximum pods limitation, 3) Maximum IP address limitation.
+
+### Compute resource limitation
+Each InferenceService has a resource overhead because of the sidecars injected into each pod. 
+This normally adds about 0.5 CPU and 0.5G Memory resource per InferenceService replica. 
+For example, if we deploy 10 models, each with 2 replicas, then the resource overhead is 10 * 2 * 0.5 = 10 CPU and 10 * 2 * 0.5 = 10 GB memory. 
+Each model’s resource overhead is 1CPU and 1 GB memory. 
+Deploying many models using the current approach will quickly use up a cluster's computing resource. 
+With Multi-model serving, these models can be loaded in one InferenceService, 
+then each model's average overhead is 0.1 CPU and 0.1GB memory.
+For GPU based models, the number of GPUs required grows linearly as the number of models grows, which is not cost efficient. 
+If multiple models can be loaded in one GPU enabled model server such as TritonServer, we need a lot less GPUs in the cluster.
+
+### Maximum pods limitation
+Kubernetes has a maximum number of pods per node with the default limit set to 110. 
+A typical 50-node cluster with default pod limit can run at most 1000 models
+ assuming each InferenceService has 4 pods on average (two transformer replicas and two predictor replicas).
+
+### Maximum IP address limitation.
+Kubernetes clusters also have an IP address limit per cluster. 
+Each pod in InferenceService needs an independent IP. 
+For example a cluster with 4096 IP addresses can deploy at most 1024 models assuming each InferenceService has 4 pods on average (two transformer replicas and two predictor replicas).
```

Comment:
Is there any reference on the number for 4096? A kubernetes cluster needs to allocate pod and service CIDR blocks, I think this is collapsed to the above point with the 110 pod limit on each node which is usually allocated with 256 IP addresses.

Reply:
The 110 limit comes from the kubelet default setting: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ and the k8s's best practice guides says that a node shouldn't run more than 100 pods: https://kubernetes.io/docs/setup/best-practices/cluster-large/

Reply:
The scalability sig wants to increase this number go 500/node: https://github.com/kubernetes/community/blob/master/sig-scalability/goals.md#derived-metricscaveats

Reply:
The number of IP address can be different per CNI used by cluster admin, the number can very a lot according to Ned. Could double check with him tomorrow.